src/List.elm

```diff
@@ -387,7 +391,12 @@ concat lists =
 -}
 concatMap : (a -> List b) -> List a -> List b
 concatMap f list =
-  concat (map f list)
+  let
+    helper val acc =
+      append (f val) acc
+  in
+    foldr helper [] list
```

Comment:
@sparksp figured out a faster `concatMap`:
```elm
fasterConcatMap : (a -> List b) -> List a -> List b
fasterConcatMap f =
    List.foldr (f >> (++)) []
```
Granted, this doesn't take the JS changes into account, so maybe overall your version is faster, but in current `elm/core` the above snippet wins the benchmark.

Reply:
Luckily this is the same as the proposed concatMap 😄 `(f >> (++))` is functionally the same as `helper` just less expressive.

Reply:
beware: `append` vs `(++)` does introduce a difference, IIRC.

Reply:
The JS changes in this PR makes `++` and `List.append` equally fast. So if this PR was to be merged, I don't think there would be a difference between these two implementations.

Reply:
exactly. The point of this PR is to make the non-obvious performance difference between `List.append` and `(++)` to go away which will implicitly improve all the other functions defined in terms of `append`.

I wonder if you have any idea how @evancz feels about the change @Skinney. It's been a while since this was opened.

Reply:
I remember that me and Evan looked through this, but I forget what the conclusion of that meeting was. I don't think he was opposed to any of the changes, though.